perf: parallelize database queries in item data processing

jorgetamayo21 opened the pull request:
- Implement parallel execution of initial database queries using Promise.all()
- Optimize chart data queries by executing all intervals concurrently
- Add conditional promise handling for APDEX and SUT metrics
- Separate database querying from data processing logic
- Reorder scenario settings loading for better conditional flow

Performance improvements tested with 55M+ records:
- 69.7% reduction in processing time (3m44s → 1m8s)
- 251% increase in throughput (247K → 867K records/sec)
- Successfully handles larger datasets with improved efficiency

Fixes sequential bottlenecks in report generation pipeline.

jorgetamayo21 committed:
perf: parallelize database queries in item data processing

- Implement parallel execution of initial database queries using Promise.all()
- Optimize chart data queries by executing all intervals concurrently
- Add conditional promise handling for APDEX and SUT metrics
- Separate database querying from data processing logic
- Reorder scenario settings loading for better conditional flow

Performance improvements tested with 55M+ records:
- 69.7% reduction in processing time (3m44s → 1m8s)
- 251% increase in throughput (247K → 867K records/sec)
- Successfully handles larger datasets with improved efficiency

Fixes sequential bottlenecks in report generation pipeline.

jorgetamayo21 commented:
@ludeknovy I've been using this change in my local implementation for 2 months now and it has worked without any problems.

ludeknovy commented:
Hi @jorgetamayo21 
Thanks! I will review it and let you know. Sorry for the delay.

ludeknovy commented on src/server/controllers/item/shared/item-data-processing.ts:
```diff
@@ -42,50 +42,77 @@ export const itemDataProcessing = async ({ projectName, scenarioName, itemId })
+            dbPromises.push(db.many(calculateApdexValues(itemId, satisfyingThreshold, toleratingThreshold)))
+        }
+        
+        const aggOverview = await aggOverviewPromise
```
can we do `const aggOverview = await db.one(aggOverviewQuery(itemId))` instead?
I find it easier to read.

jorgetamayo21 replied:
applied in the new commit

ludeknovy commented on src/server/controllers/item/shared/item-data-processing.ts:
```diff
@@ -42,50 +42,77 @@ export const itemDataProcessing = async ({ projectName, scenarioName, itemId })
         }
 
         if (scenarioSettings.apdexSettings.enabled) {
+            apdex = dbResults[conditionalIndex++]
```
This seems to be a little tricky, if a new query is added as a last element to the array, this logic gets broken?

Would not this be better?

```
let dbPromises = [
            db.many(aggLabelQuery(itemId)),
            db.manyOrNone(responseCodeDistribution(itemId)),
            db.manyOrNone(responseTimePerLabelHistogram(itemId)),
            db.manyOrNone(responseMessageFailures(itemId)),
            db.manyOrNone(getDownsampledRawData(itemId, MAX_SCATTER_CHART_POINTS)),
            db.manyOrNone(findGroupedErrors(itemId)),
            db.manyOrNone(findTop5ErrorsByLabel(itemId))
        ]
```

and 

```
const [
  aggLabels,
  responseCodeDistribution,
  responseTimeHistogram,
  responseMessageFailures,
  downsampledRawData,
  groupedErrors,
  top5ErrorsByLabel
] = await Promise.all(Object.values(dbPromises));
```

with conditional logic for the apdex query based on the `scenarioSettings.apdexSettings.enabled` ?

jorgetamayo21 replied:
You're right, I separated apdex from the array and now it's handled independently.

It was applied in the new commit.

jorgetamayo21 committed:
fix: simplify aggOverview query and handle apdex independently in item-data-processing.

jorgetamayo21 commented:
I haven't been able to test the new commit yet, but I'll let you know when I can.

jorgetamayo21 commented:
> I haven't been able to test the new commit yet, but I'll let you know when I can.


Last commit tested with 58 million records processed in 1 minute and 44 seconds

be-1  | {"level":"info","message":"Stopping async item: e0fbd847-7062-4675-b129-3ecd85e14f0e","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading scenario settings","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading label aggregation","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading status code distribution","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading response time per label distribution","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading response failures","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading raw downsampled data","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading grouped errors","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading top 5 errors by label","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"Loading overview aggregation","timestamp":"2026-02-26 15:40:42"}
be-1  | {"level":"debug","message":"User d01dc8a1-203a-423b-995a-98c527d1a191 with role admin accessing a resource with allowed roles: readonly,operator,admin","timestamp":"2026-02-26 15:41:05"}
be-1  | {"level":"info","message":"GET /api/projects/CORP/scenarios/perf-locust-test/processing-items?environment= 304 38ms","meta":{}}
be-1  | {"level":"debug","message":"Preparing queries for interval: 14424 milliseconds","timestamp":"2026-02-26 15:41:40"}
be-1  | {"level":"debug","message":"Executing all chart queries in parallel","timestamp":"2026-02-26 15:41:40"}
be-1  | {"level":"debug","message":"User d01dc8a1-203a-423b-995a-98c527d1a191 with role admin accessing a resource with allowed roles: readonly,operator,admin","timestamp":"2026-02-26 15:42:05"}
be-1  | {"level":"info","message":"GET /api/projects/CORP/scenarios/perf-locust-test/processing-items?environment= 304 6ms","meta":{}}
be-1  | {"level":"debug","message":"Processing data for interval: 14424 milliseconds","timestamp":"2026-02-26 15:42:26"}
be-1  | {"level":"info","message":"Item: e0fbd847-7062-4675-b129-3ecd85e14f0e processing finished","timestamp":"2026-02-26 15:42:26"}

jorgetamayo21 commented:
Hi @ludeknovy , I've been using this implementation for months and haven't had any problems. I'm tagging you so I don't get forgotten, haha.

ludeknovy commented:
@jorgetamayo21 
Thanks. Gonna check it.

ludeknovy committed:
Merge branch 'master' into master

ludeknovy commented:
@jorgetamayo21 there are some linter issues. 
Please run npm run lint:fix. It will fix most of the problems. But as copilot pointed out, there are some unused variables now as well that cannot be fixed automatically.

jorgetamayo21 committed:
fix: resolve unhandled rejections, pool saturation and unused vars

- Await aggOverview before building dbPromises to eliminate unhandled rejection window
- Fold sutMetrics and apdex queries into a single Promise.all via index tracking,
  removing the separate sutMetricsPromise that could reject without a handler
- Introduce queueItemProcessing to serialize item processing and prevent DB
  connection pool saturation under concurrent uploads, logging errors on queue failure
- Apply queueItemProcessing in both create and stop-async controllers
- Remove unused variables distributedThreads and sutMetricsPromise
- Replace Object.entries loop with direct array iteration to drop unused index var

jorgetamayo21 commented:
@ludeknovy  All the review observations have been addressed. Additionally, I introduced a FIFO queue to serialize item processing and prevent DB connection pool saturation under concurrent processing.

jorgetamayo21 committed:
Merge branch 'ludeknovy:master' into master
